src/main/java/com/vlkan/pubsub/PubsubSubscriber.java

```diff
@@ -0,0 +1,222 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permits and
+ * limitations under the License.
+ */
+
+package com.vlkan.pubsub;
+
+import com.vlkan.pubsub.model.PubsubAckRequest;
+import com.vlkan.pubsub.model.PubsubPullRequest;
+import com.vlkan.pubsub.model.PubsubPullResponse;
+import com.vlkan.pubsub.model.PubsubReceivedAckableMessage;
+import com.vlkan.pubsub.ratelimiter.StagedRateLimiter;
+import com.vlkan.pubsub.util.FluxHelpers;
+import io.micrometer.core.instrument.DistributionSummary;
+import io.micrometer.core.instrument.MeterRegistry;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+import reactor.core.scheduler.Scheduler;
+
+import java.math.BigDecimal;
+import java.time.Duration;
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+public class PubsubSubscriber {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(PubsubSubscriber.class);
+
+    private final PubsubSubscriberConfig config;
+
+    private final PubsubClient client;
+
+    private final Scheduler scheduler;
+
+    private final PullResponseConsumer pullResponseConsumer;
+
+    private final PubsubPullRequest pullRequest;
+
+    private final StagedRateLimiter rateLimiter;
+
+    private final DistributionSummary permitWaitPeriodDistributions;
+
+    @FunctionalInterface
+    public interface PullResponseConsumer {
+
+        Mono<PubsubAckRequest> accept(PubsubPullResponse pullResponse);
+
+    }
+
+    private PubsubSubscriber(Builder builder) {
+        this.config = builder.config;
+        this.client = builder.client;
+        this.scheduler = builder.scheduler;
+        this.pullResponseConsumer = builder.pullResponseConsumer;
+        this.pullRequest = new PubsubPullRequest(true, config.getPullBufferSize());
+        String rateLimiterName = String.format("%s/%s", config.getProjectName(), config.getSubscriptionName());
+        this.rateLimiter = StagedRateLimiter.of(rateLimiterName, config.getRateLimiterSpec());
+        this.permitWaitPeriodDistributions = DistributionSummary
+                .builder(config.getMeterName())
+                .register(builder.meterRegistry);
+    }
+
+    public PubsubSubscriberConfig getConfig() {
+        return config;
+    }
+
+    public PubsubClient getClient() {
+        return client;
+    }
+
+    public Scheduler getScheduler() {
+        return scheduler;
+    }
+
+    public PullResponseConsumer getPullResponseConsumer() {
+        return pullResponseConsumer;
+    }
+
+    public Flux<PubsubSubscriberConsumption> start() {
+        return Flux
+                .just(new AtomicBoolean(false))
```

Comment:
could you please clarify the use of `AtomicBoolean` here? I have a feeling that it may not work as you expect it to work, but need to understand it better first

Reply:
I use it as a flag holder indicating if the previous pull returned an empty batch. If so, retry in `PubsubSubscriberConfig#pullPeriod`; otherwise, retry immediately. What do you think about this model in general? I find it utterly simply and working pretty effectively according to my experience.

Another approach I could have taken could be using `PubsubPullRequest`s where `immediateReturnEnabled=false`. Then Pub/Sub server puts the request on hold until a message arrives. Though I refrained from this approach due to the following concerns:

- Even the official Pub/Sub driver uses `immediateReturnEnabled=true`.
- `immediateReturnEnabled=true` enforces a deterministic model. There is no unknown wait period. Though `immediateReturnEnabled=false` totally makes the lifetime of the request depend on the emotional state of Pub/Sub servers, spoils the deterministic nature of the anticipated behavior, pollutes the metrics with random wait periods, and I don't know how well that will play along with `PubsubClientConfig#pullPeriod`-based timeouts.

Reply:
Have you considered using `repeatWhenEmpty`?

Reply:
I did not understand. Would you mind elaborating on the idea a little bit more, please? Do you mean repeating-`Mono<PubsubPullResponse>`-*unless*-it's-empty?

Reply:
If I understand the logic correctly, you pull again with a delay if the response was empty, and the `AtomicBoolean` is used for it.

However, you can use `repeatWhenEmpty` operator to achieve the same, here is a pseudo code that demonstrates the idea:
```
var pullResponses = pull()
    .repeatWhenEmpty(it -> it.delayElements(intervalBetweenEmptyPulls))
    .repeat();
```

Reply:
Sweet!

```java
public Flux<PubsubPullResponse> pullAll() {
    return client
            .pull(config.getProjectName(), config.getSubscriptionName(), pullRequest)
            .filter(pullResponse -> !pullResponse.getReceivedAckableMessages().isEmpty())
            .repeatWhenEmpty(flux -> scheduler == null
                    ? flux.delayElements(config.getPullPeriod())
                    : flux.delayElements(config.getPullPeriod(), scheduler))
            .repeat();
}
```

Reply:
Told ya ;)

P.S. Since `scheduler` variable is immutable, you can also do this:
```java
.repeatWhenEmpty(
        scheduler == null
                    ? flux -> flux.delayElements(config.getPullPeriod())
                    : flux -> flux.delayElements(config.getPullPeriod(), scheduler)
)
```